openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	pod.Annotations[workloadAnnotation] = fmt.Sprintf(`{"%s":"%s"}`, podWorkloadAnnotationEffect, workloadEffectPreferredDuringScheduling)
+}
+
+func updateContainersResources(containers []coreapi.Container, podAnnotations map[string]string, workloadType string, podQoSClass coreapi.PodQOSClass) {
+	for i := range containers {
+		c := &containers[i]
+		cpusAnnotationKey := fmt.Sprintf("%s/%s", containerCPUSharesAnnotation, c.Name)
+
+		if podQoSClass == coreapi.PodQOSBestEffort {
+			podAnnotations[cpusAnnotationKey] = "2"
+			continue
+		}
+
+		if c.Resources.Requests != nil {
+			// if the kubelet deleted container CPUs requests from resources for static pods,
+			// the admission plugin will not do additional updates
+			if _, ok := c.Resources.Requests[coreapi.ResourceCPU]; !ok {
+				continue
+			}
+
+			cpuRequest := c.Resources.Requests[coreapi.ResourceCPU]
+			cpuRequestInMilli := cpuRequest.MilliValue()
+
+			cpuShares := cm.MilliCPUToShares(cpuRequestInMilli)
+			podAnnotations[cpusAnnotationKey] = fmt.Sprintf("%d", cpuShares)
+			delete(c.Resources.Requests, coreapi.ResourceCPU)
+
+			if c.Resources.Limits == nil {
+				c.Resources.Limits = coreapi.ResourceList{}
+			}
+
+			// multiple the CPU request by 1000, to make sure that the resource will pass integer validation
+			managedResource := fmt.Sprintf("%s.%scores", workloadType, podWorkloadAnnotationPrefix)
+			newCPURequest := resource.NewMilliQuantity(cpuRequestInMilli*1000, cpuRequest.Format)
+			c.Resources.Requests[coreapi.ResourceName(managedResource)] = *newCPURequest
+			c.Resources.Limits[coreapi.ResourceName(managedResource)] = *newCPURequest
+		}
+	}
+}
+
+func isGuaranteed(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request, requestExist := c.Resources.Requests[r]
+
+			if limit.IsZero() {
+				return false
+			}
+
+			if !requestExist {
+				continue
+			}
+
+			// it some corner case, when you set CPU request to 0 the k8s will change it to the value
+			// specified under the limit
+			if r == coreapi.ResourceCPU && request.IsZero() {
+				continue
+			}
+
+			if !limit.Equal(request) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isBestEffort(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request := c.Resources.Requests[r]
+
+			if !limit.IsZero() || !request.IsZero() {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getPodQoSClass(containers []coreapi.Container) coreapi.PodQOSClass {
```

Comment:
why not use the upstream func pkg/apis/core/v1/helper/qos/qos.go:39 ?

Reply:
Is this now the complete logic we have upstream or still some incomplete clone?

Reply:
IIUC the function under `pkg/apis/core/v1/helper/qos/qos.go` is assuming that it got pod after the defaulter when in case of the admission plugin is does not true(please correct me if I wrong)
As result, it returns Buarstable class for the unit test:
```
{
	name:             "should recognize guaranteed pod when CPU request equals to 0",
	pod:              testManagedPod("100m", "0", "100Mi", "100Mi"), // cpuLimit, cpuRequest, memoryLimit, memoryRequest
	expectedQoSClass: coreapi.PodQOSGuaranteed,
},
```
If you insist on the usage of the method from `pkg/apis/core/v1/helper/qos/qos.go` I can somehow change and use it under the admission plugin(not too optimistic about it).